pkg/reconciler/ingress/resources/virtual_service.go

```diff
@@ -0,0 +1,257 @@
+	"github.com/knative/serving/pkg/apis/networking/v1alpha1"
+	"github.com/knative/serving/pkg/apis/serving"
+	"github.com/knative/serving/pkg/network"
+	"github.com/knative/serving/pkg/reconciler/ingress/resources/names"
+	"github.com/knative/serving/pkg/resources"
+)
+
+// VirtualServiceNamespace gives the namespace of the child
+// VirtualServices for a given Ingress.
+func VirtualServiceNamespace(_ *v1alpha1.Ingress) string {
+	return system.Namespace()
+}
+
+// MakeIngressVirtualService creates Istio VirtualService as network
+// programming for Istio Gateways other than 'mesh'.
+func MakeIngressVirtualService(ci *v1alpha1.Ingress, gateways []string) *v1alpha3.VirtualService {
+	vs := &v1alpha3.VirtualService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            names.IngressVirtualService(ci),
+			Namespace:       VirtualServiceNamespace(ci),
+			OwnerReferences: []metav1.OwnerReference{*kmeta.NewControllerRef(ci)},
+			Annotations:     ci.ObjectMeta.Annotations,
+		},
+		Spec: *makeVirtualServiceSpec(ci, gateways, expandedHosts(getHosts(ci))),
+	}
+
+	// Populate the Ingress labels.
+	if vs.Labels == nil {
+		vs.Labels = make(map[string]string)
+	}
+	vs.Labels[networking.IngressNamespaceLabelKey] = ci.Namespace
+	vs.Labels[networking.IngressLabelKey] = ci.Name
+
+	ingressLabels := ci.Labels
+	vs.Labels[serving.RouteLabelKey] = ingressLabels[serving.RouteLabelKey]
+	vs.Labels[serving.RouteNamespaceLabelKey] = ingressLabels[serving.RouteNamespaceLabelKey]
+	return vs
+}
+
+// MakeMeshVirtualService creates Istio VirtualService as network
+// programming for Istio network mesh.
+func MakeMeshVirtualService(ci *v1alpha1.Ingress) *v1alpha3.VirtualService {
+	vs := &v1alpha3.VirtualService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            names.MeshVirtualService(ci),
+			Namespace:       VirtualServiceNamespace(ci),
+			OwnerReferences: []metav1.OwnerReference{*kmeta.NewControllerRef(ci)},
+			Annotations:     ci.ObjectMeta.Annotations,
+		},
+		Spec: *makeVirtualServiceSpec(ci, []string{"mesh"}, retainLocals(getHosts(ci))),
+	}
+	// Populate the Ingress labels.
+	vs.Labels = resources.UnionMaps(
+		resources.FilterMap(ci.Labels, func(k string) bool {
+			return k != serving.RouteLabelKey && k != serving.RouteNamespaceLabelKey
+		}),
+		map[string]string{networking.IngressLabelKey: ci.Name,
+			networking.IngressNamespaceLabelKey: ci.Namespace})
+	return vs
+}
+
+// MakeVirtualServices creates Istio VirtualServices as network programming.
+//
+// These VirtualService specifies which Gateways and Hosts that it applies to,
+// as well as the routing rules.
+func MakeVirtualServices(ci *v1alpha1.Ingress, gateways []string) []*v1alpha3.VirtualService {
+	vss := []*v1alpha3.VirtualService{MakeMeshVirtualService(ci)}
+	if len(gateways) > 0 {
+		vss = append(vss, MakeIngressVirtualService(ci, gateways))
+	}
+	return vss
+}
+
+func makeVirtualServiceSpec(ci *v1alpha1.Ingress, gateways []string, hosts []string) *v1alpha3.VirtualServiceSpec {
```

Comment:
I believe this only depends on `ci.Spec`, which could be shared with ClusterIngress controller.

During migration we rely on having the VirtualServices owned by namespaced Ingress and ClusterIngress to be identical, so code sharing here is important to avoid discrepancies.

Reply:
@mattmoor are you against having code in `pkg/reconciler/ingress/resources/virtual_service.go` referring to `pkg/reconciler/clusteringress/resources/...` ?

Reply:
Resources have been refactored and shared by iingress/clusteringress controllers.

Reply:
See my comment about having these in their final resting place, otherwise sharing SGTM.